File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    #today를 기준
    #약관 별 term
    #privacies는 약관 가입 날짜
    #privacies 가입 날짜에서 term을 더해서 today 넘지 않았으면 answer.append() 
    #=> if privacies+term<=today : answer.append()
    answer = []
    term_dict={}
    today_year,today_month,today_day=today.split(".")
    today_year,today_month,today_day=int(today_year),int(today_month),int(today_day)
    count=1
    for i in terms:
        fir,sec=i.split()
        term_dict[fir]=int(sec)

    for i in privacies:
        #day와 약관 나눔
        day,category=i.split()
        #년,월,일 나눔
        year,month,day=day.split('.')
        year,month,day=int(year),int(month),int(day)
        month+=term_dict[category]
        if month>12:
            year+=month//12
            if month % 12 == 0:
                year -= 1
                month = 12
            else:
                month %= 12

        #기간 넘은거 answer에 추가
        if today_year>year:
            answer.append(count)
        elif today_year==year and today_month>month:
            answer.append(count)
        elif today_year==year and today_month==month and today_day>=day:
            answer.append(count)
        count+=1

    return answer

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;

vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    map<char, int> mp;

    // 현재 날짜
    int year = stoi(today.substr(0, today.find(".")));
    today.erase(0, today.find(".") + 1);
    int month = stoi(today.substr(0, today.find(".")));
    today.erase(0, today.find(".") + 1);
    int day = stoi(today);


    // 맵에 약관종류 넣기
    for (string str : terms)
    {
        int pos = str.find(" ");
        char c = str[0];
        int m = stoi(str.substr(pos, str.size() - 1));
        mp[c] = m;
    }

    int idx = 1;
    for (string str : privacies)
    {
        vector<int> v;
        int pos = 0;

        // 년월 int로 넣기
        while ((pos = str.find(".")) != string::npos)
        {
            int a = stoi(str.substr(0, pos));
            v.push_back(a);
            str.erase(0, pos + 1);
        }
        // 일 넣기
        int d = stoi(str.substr(0, str.find(" ")));
        v.push_back(d);
        str.erase(0, str.find(" ") + 1);

        // 계약 기간에 해당하는 만큼 날짜 변경하기
        int m = mp[str[0]];
        // 달에 추가
        v[1] += m;

        int plus = v[1] / 12;
        int remain = v[1] % 12;
        if (remain == 0) {
            plus--;
            remain = 12;
        }
        v[0] += plus;
        v[1] = remain;

        // 현재 날짜를 기준으로 년도 비교 월 비교 일 비교하기
        if (v[0] < year)
            answer.push_back(idx);
        else if (v[0] == year && v[1] < month)
            answer.push_back(idx);
        else if (v[0] == year && v[1] == month && v[2] - 1 < day)
            answer.push_back(idx);

        idx++;
    }

    // 개인정보를 기준으로 약관 지난 기간으로 작성



    return answer;
}